src/input/input_python.rs

```diff
@@ -592,6 +594,25 @@ impl<'py> Input<'py> for Bound<'py, PyAny> {
 
         Err(ValError::new(ErrorTypeDefaults::TimeDeltaType, self))
     }
+
+    fn validate_complex<'a>(&'a self, _py: Python<'py>) -> ValResult<ValidationMatch<EitherComplex<'py>>> {
+        if let Ok(complex) = self.downcast::<PyComplex>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Py(complex.to_owned())));
+        } else if let Ok(s) = self.downcast::<PyString>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Py(string_to_complex(s, self)?)));
+        } else if self.is_exact_instance_of::<PyFloat>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Complex([
+                self.extract::<f64>().unwrap(),
+                0.0,
+            ])));
+        } else if self.is_exact_instance_of::<PyInt>() {
+            return Ok(ValidationMatch::exact(EitherComplex::Complex([
+                self.extract::<i64>().unwrap() as f64,
+                0.0,
+            ])));
+        }
+        Err(ValError::new(ErrorTypeDefaults::ComplexType, self))
+    }
```

Comment:
We should have a think here about what behaviour we expect in `strict` mode. Probably we only want to accept `complex` objects in strict. Do we want to accept integers / floats in "strict" mode? In lax, we can accept everything for sure.

Also, similarly the `ValidationMatch::exact` calls should probably be downgraded to either `strict` or `lax` - these will set the level of priority in e.g. `complex | int` union. I think most would agree that `complex | int` would expect an integer value to _not_ be cast to a complex. We might want some according tests for some of these union cases.

Reply:
`strict` mode is strict in terms of typing, right? I was mostly thinking from the perspective of mathematics when I added `float` and `int`. I suppose then we should only allow `complex` objects in strict mode? Then `ValidationMatch::exact` is the same as `strict`. This sounds reasonable to me, but I'm not exactly sure about the how different `exact` and `strict` usually are.

One more question about `strict` mode: how does it apply to other input types? Is it like when the input is in JSON, we only accept strings in `strict` mode?

Reply:
> I suppose then we should only allow `complex` objects in strict mode?

I think so, yes (note that in JSON it's less straightforward as you note). In JSON I think we only accept strings in `strict` mode, yes.

> Then `ValidationMatch::exact` is the same as `strict`.

IIRC in other cases I made subclasses `strict`, if `complex` objects allow subclassing. I'd check what I did for ints or floats.